src/xds.c

```diff
@@ -25,38 +25,154 @@
 #include "xds.h"
 #include "caption.h"
 #include <string.h>
+#include <xds.h>
+#include <caption.h>
+#include <xds_data.h>
+#include <stdio.h>
+
+int is_valid_type(uint8_t class_code, uint8_t type) {
+    if (type == 0) {
+        return 0;
+    }
+    switch(class_code) {
+        case CURRENT: // valid types are 0x01 to 0x17 (ANSI-CTA-608-E R-2014 9.5.1)
+            return type <= 0x17;
+        case FUTURE: // valid types are 0x01 to 0x17 (ANSI-CTA-608-E R-2014 9.5.2)
+            return type <= 0x17;
+        case CHANNEL: // valid types are 0x01 to 0x04 (ANSI-CTA-608-E R-2014 9.5.3)
+            return type <= 0x04;
+        case MISC: // valid types are 0x01 to 0x04, and 0x40 to 0x43 (ANSI-CTA-608-E R-2014 9.5.4)
+            return type <= 0x04 || (type >= 0x40 && type <= 0x43);
+        case PUBLIC_SERVICE: // valid types are 0x01 to 0x02 (ANSI-CTA-608-E R-2014 9.5.5)
+            return type <= 0x02;
+        case CLASS_RESERVED: // used arbitrarily by xds encoding equipment (ANSI-CTA-608-E R-2014 9.5)
+            return 1;
+        case PRIVATE_DATA: // used arbitrarily (ANSI-CTA-608-E R-2014 9.6)
+            return 1;
+        default:
+            return 0;
+    }
+}
+
+void clear_packet_info(xds_packet_t* packet) {
+    memset(packet, 0, sizeof(xds_packet_t));
+}
 
 void xds_init(xds_t* xds)
 {
     memset(xds, 0, sizeof(xds_t));
 }
 
-int xds_decode(xds_t* xds, uint16_t cc)
+libcaption_status_t xds_decode(caption_frame_t* frame, uint16_t cc)
 {
+    xds_t* xds = &frame->xds;
+
     switch (xds->state) {
-    default:
-    case 0:
-        xds_init(xds);
-        xds->class_code = (cc & 0x0F00) >> 8;
-        xds->type = (cc & 0x000F);
-        xds->state = 1;
-        return LIBCAPTION_OK;
-
-    case 1:
-        if (0x8F00 == (cc & 0xFF00)) {
-            xds->checksum = (cc & 0x007F);
-            xds->state = 0;
-            return LIBCAPTION_READY;
-        }
-
-        if (xds->size < 32) {
-            xds->content[xds->size + 0] = (cc & 0x7F00) >> 8;
-            xds->content[xds->size + 1] = (cc & 0x007F);
-            xds->size += 2;
+        default:
+        case 0: // A control code was seen after a non-XDS data stream
+            ;
```

Comment:
is this intended?

Reply:
yea, in C the first statement after a label can't be a variable declaration, so you need an empty statement (;) if your first line is a variable declaration

Reply:
I'm surprised by this...

Reply:
Other way to work around this is to set the scope {
}

Reply:
I can change it to that if you'd like, I figured it was more consistent with an empty statement because none of the other cases have curly braces